lib/concentrate/supervisor.ex

```diff
@@ -27,108 +20,9 @@ defmodule Concentrate.Supervisor do
   end
 
   def children(opts) do
-    {source_names, source_children} = sources(opts)
-
-    merge = merge(source_names)
-
-    consumers = consumers()
-
-    Enum.concat([source_children, merge, consumers])
-  end
-
-  def sources(opts) do
-    realtime_enhanced_child =
-      if opts[:busloc_url] do
-        source_child(
-          :gtfs_realtime_enhanced,
-          opts[:busloc_url],
-          Concentrate.Parser.GTFSRealtimeEnhanced
-        )
-      else
-        nil
-      end
-
-    swiftly_child =
-      if opts[:swiftly_realtime_vehicles_url] && opts[:swiftly_authorization_key] do
-        source_child(
-          :swiftly_realtime_vehicles,
-          opts[:swiftly_realtime_vehicles_url],
-          Concentrate.Parser.SwiftlyRealtimeVehicles,
-          headers: %{
-            "Authorization" => opts[:swiftly_authorization_key]
-          },
-          params: %{
-            unassigned: "true",
-            verbose: "true"
-          }
-        )
-      else
-        nil
-      end
-
-    trip_updates_child =
-      if opts[:trip_updates_url] do
-        source_child(
-          :trip_updates_enhanced,
-          opts[:trip_updates_url],
-          Concentrate.Parser.GTFSRealtimeEnhanced
-        )
-      else
-        nil
-      end
-
-    children =
-      [realtime_enhanced_child, swiftly_child, trip_updates_child]
-      |> Enum.reject(&is_nil/1)
-
-    {child_ids(children), children}
-  end
-
-  defp source_child(source, url, parser, opts \\ []) do
-    Supervisor.child_spec(
-      {
-        Concentrate.Producer.HTTP,
-        {url, [name: source, parser: parser] ++ opts}
-      },
-      id: source
-    )
-  end
-
-  def merge(source_names) do
-    sources = outputs_with_options(source_names, max_demand: 1)
-
-    [
-      {
-        Concentrate.Merge,
-        name: :merge, subscribe_to: sources, buffer_size: 1
-      }
-    ]
-  end
-
-  def consumers do
-    vehicle_positions_consumer =
-      consumer(Concentrate.Consumer.VehiclePositions, :vehicle_positions)
-
-    stop_time_updates_consumer =
-      consumer(Concentrate.Consumer.StopTimeUpdates, :stop_time_updates)
-
-    [vehicle_positions_consumer, stop_time_updates_consumer]
-  end
-
-  def consumer(module, id) do
-    Supervisor.child_spec(
-      {module, subscribe_to: [merge: [max_demand: 1]]},
-      id: id
-    )
-  end
-
-  defp child_ids(children) do
-    for child <- children, do: child.id
-  end
-
-  def outputs_with_options(outputs, options) do
-    for name <- outputs do
-      {name, options}
-    end
+    Enum.concat([
+      Concentrate.Pipeline.VehiclePositionsPipeline.pipeline(opts),
+      Concentrate.Pipeline.StopTimeUpdatesPipeline.pipeline(opts)
+    ])
```

Comment:
I'm thinking of the stop times pipeline as being pretty independent from the concentrate pipeline. What do you think about giving them separate supervisors to reflect that, and so if one pipeline dies, the other one doesn't have to get restarted with it. (I'm imagining both pipeline-supervisors would be started by Realtime.Supervisor)

Reply:
Added individual supervisors for each pipeline.

Reply:
👍 They can be restarted separately, but I think the structure is more complex than necessary.
It's now
```
Realtime.Supervisor
├ ...
└ Concentrate.Supervisor
  ├ Concentrate.Pipeline.VehiclePositionsPipelineSupervisor
  │ └ Concentrate.Pipeline.VehiclePositionsPipeline
  │   ├ Concentrate.Producer.HTTP
  │   ├ Concentrate.Producer.HTTP
  │   ├ Concentrate.Merge
  │   └ Concentrate.Consumer.VehiclePositions
  └ Concentrate.Pipeline.StopTimeUpdatesPipelineSupervisor
    └ Concentrate.Pipeline.StopTimeUpdatesPipeline
      ├ Concentrate.Producer.HTTP
      └ Concentrate.Consumer.StopTimeUpdates
```
but we can knock out a couple layers like this:
```
Realtime.Supervisor
├ ...
├ Concentrate.VehiclePositionsPipeline
│ ├ Concentrate.Producer.HTTP
│ ├ Concentrate.Producer.HTTP
│ ├ Concentrate.Merge
│ └ Concentrate.Consumer.VehiclePositions
└ Concentrate.Pipeline.StopTimeUpdatesPipeline
  ├ Concentrate.Producer.HTTP
  └ Concentrate.Consumer.StopTimeUpdates
```

Reply:
That was intentional to encapsulate all the pipeline configuration knowledge. It didn't seem like something the Realtime supervisor should be concerned with.

Reply:
`Realtime.Supervisor` wouldn't know anything about how the pipelines are set up. It would just know that there's 2 separate pipelines, one for VPs and one for STUs.

All the knowledge about which children to start and which env vars to use would be written in `VehiclePositionsPipeline` or `StopTimeUpdatesPipeline`.

Reply:
Yes, knowing about what pipelines exist is what I wanted to encapsulate.